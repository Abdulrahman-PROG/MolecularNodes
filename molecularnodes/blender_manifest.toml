schema_version = "1.0.0"

# Example of manifest file for a Blender extension
# Change the values according to your extension
id = "molecularnodes"
version = "4.2.0"
name = "molecularnodes"
tagline = "Import and animation of molecular data."
maintainer = "Brady Johnston <brady.johnston@me.com>"
type = "add-on"

# Optional: add-ons can list which resources they will require:
# * "files" (for access of any filesystem operations)
# * "network" (for internet access)
# * "clipboard" (to read and/or write the system clipboard)
# * "camera" (to capture photos and videos)
# * "microphone" (to capture audio)
permissions = [
	"files",
	"network",
]

# Optional link to documentation, support, source files, etc
website = "http://bradyajohnston.github.io/MolecularNodes"

# Optional list defined by Blender and server, see:
# https://docs.blender.org/manual/en/dev/extensions/tags.html
tags = [
	"3D View",
	"Animation",
	"Node",
]

blender_version_min = "4.2.0"
# Optional: maximum supported Blender version
# blender_version_max = "5.1.0"

# License conforming to https://spdx.org/licenses/ (use "SPDX: prefix)
# https://docs.blender.org/manual/en/dev/extensions/licenses.html
license = [
  "SPDX:GPL-2.0-or-later",
]
# Optional: required by some licenses.
# copyright = [
#   "2002-2024 Developer Name",
#   "1998 Company Name",
# ]

# Optional: bundle 3rd party Python modules.
wheels = [
	"./wheels/pandas-2.2.2-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/urllib3-2.2.1-py3-none-any.whl",
	"./wheels/python_dateutil-2.9.0.post0-py2.py3-none-any.whl",
	"./wheels/starfile-0.5.6-py3-none-any.whl",
	"./wheels/typing_extensions-4.11.0-py3-none-any.whl",
	"./wheels/pyarrow-16.0.0-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/numpy-1.26.4-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/biotite-0.40.0-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/requests-2.31.0-py3-none-any.whl",
	"./wheels/pytz-2024.1-py2.py3-none-any.whl",
	"./wheels/msgpack-1.0.8-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/tzdata-2024.1-py2.py3-none-any.whl",
	"./wheels/idna-3.7-py3-none-any.whl",
	"./wheels/six-1.16.0-py2.py3-none-any.whl",
	"./wheels/mrcfile-1.4.3-py2.py3-none-any.whl",
	"./wheels/certifi-2024.2.2-py3-none-any.whl",
	"./wheels/networkx-3.3-py3-none-any.whl",
	"./wheels/charset_normalizer-3.3.2-cp311-cp311-macosx_11_0_arm64.whl",
]
