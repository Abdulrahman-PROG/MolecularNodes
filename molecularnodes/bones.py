import bpy
import numpy as np
from . import obj

def add_bones(object):
    ## creates bones and assigns correct weights
    
    bone_positions = get_bone_positions(object)
    # weights = get_bone_weights()
    
    bones = create_bones(bone_positions)
    # assign_bone_weights(object, bones, weights)
    
    return bones

def get_bone_positions(object):
    
    positions, atom_name, chain_id, res_id, sec_struct = [
        obj.get_attribute(object, att) 
        for att in ['position', 'atom_name', 'chain_id', 'res_id', 'sec_struct']
        ]
    
    sec_struct_change = np.where(np.diff(sec_struct))[0]

    pos_idx = np.where(sec_struct_change)[0]
    bone_groups = np.cumsum(sec_struct_change)
    # bone_pos_a = positions[pos_idx - 1 , :]
    bone_positions = positions[ sec_struct_change, :]
    
    
    
    
    
    print("get_bone_positions")
    return bone_positions

def get_bone_weights(object):
    print("get bone weights")

def create_bones(positions):
    
    bpy.ops.object.add(type='ARMATURE', enter_editmode=True)
    object = bpy.context.object
    object.name = 'mn_armature'
    armature = object.data
    armature.name = 'mn_frame'
    arm_name = armature.name
    bones = []
    # add bones
    for i, position in enumerate(positions):
        if i >=  len(positions) + 1:
            continue
        bone_name = f"mn_armature_{i}"
        bone = armature.edit_bones.new(bone_name)
        bone.head = position
        try:
            bone.tail = positions[i + 1, :]
        except:
            pass
        bones.append(bone.name)
    
    bpy.ops.object.editmode_toggle()
    bpy.ops.object.editmode_toggle()

    armature = bpy.data.armatures[arm_name]
    bones_a = bones.copy()
    bones_b = bones.copy()
    bones_b.pop(0)
    bones = zip(bones_a, bones_b)
    # print(f"{list(bones)=}")
    for bone_a, bone_b in bones:
        print(f"{bone_a=}")
        print(f"{bone_b=}")
        armature.edit_bones.active = armature.edit_bones[bone_a]
        for bone in [bone_a, bone_b]:
            armature.edit_bones[bone].select = True
        bpy.ops.armature.parent_set(type='CONNECTED')
        for bone in [bone_a, bone_b]:
            armature.edit_bones[bone].select = False

    print("create boens")

def assign_bone_weights(object, bones, weights):
    print("assigning bone weights")
