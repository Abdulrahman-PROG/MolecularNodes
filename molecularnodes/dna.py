import numpy as np
import bpy
from . import obj
from . import coll

bpy.types.Scene.MN_import_oxdna_top = bpy.props.StringProperty(
    name = 'Toplogy File', 
    description = 'File path for the star file to import.', 
    options = {'TEXTEDIT_UPDATE'}, 
    default = '/Users/brady/git/origami/oxdna-viewer/examples/icosahedron/7_icosahedron.json.top', 
    subtype = 'FILE_PATH', 
    maxlen = 0
    )
bpy.types.Scene.MN_import_oxdna_oxdna = bpy.props.StringProperty(
    name = 'oxDNA File', 
    description = 'File path for the oxDNA file to import.', 
    options = {'TEXTEDIT_UPDATE'}, 
    default = '/Users/brady/git/origami/oxdna-viewer/examples/icosahedron/7_icosahedron.json_post_dynamics.oxdna', 
    subtype = 'FILE_PATH', 
    maxlen = 0
    )
bpy.types.Scene.MN_import_oxdna_name = bpy.props.StringProperty(
    name = 'DNA Name', 
    description = 'Name of the created object.', 
    options = {'TEXTEDIT_UPDATE'}, 
    default = 'oxDNA', 
    subtype = 'NONE', 
    maxlen = 0
    )


def read_oxdna(filepath):
    # Open the file and read its contents
    with open(filepath, 'r') as file:
        contents = file.read()

    # Split the contents into lines
    lines = np.array(contents.split('\n'))
    is_meta = np.char.find(lines, '=') > 0

    group_id = np.cumsum(np.append([True], np.diff(is_meta)))
    groups = np.unique(group_id)
    
    frames = []

    for group in groups:
        mask = group == group_id
        if "=" in lines[mask][0]:
            continue
        
        arr = np.loadtxt(lines[mask])
        frames.append(arr)
    
    return np.stack(frames)


def read_top(filepath):
    with open(filepath, 'r') as file:
        contents = file.read()

    lines = np.array(contents.split('\n'))
    meta = lines[0]
    arr = np.loadtxt(lines[1:], dtype=str)
    # convert the columns to numeric
    arr_numeric = np.zeros(arr.shape, dtype = int)
    arr_numeric[:, (0, 2, 3)] = arr[:, (0, 2, 3)].astype(int)
    arr_numeric[:, 1] = np.unique(arr[:, 1], return_inverse=True)[1] + 30 # add offset for numeric rep of bases
    # arr_numeric[arr_numeric == -1] = arr_numeric.shape[0]
    
    return arr_numeric

def add_attributes_to_dna_mol(mol, frame, dna_scale = 0.1):
    attributes = ('base_vector', 'base_normal', 'velocity', 'angular_velocity')
    for i, att in enumerate(attributes):
        col_idx = np.array([3, 4, 5]) + i * 3
        
        try:
            data = frame[:, col_idx]
        except IndexError as e:
            print(f"Unable to get {att} attribute from coordinates. Error: {e}")
            continue
        
        if att != "angular_velocity":
            data *= dna_scale
        
        obj.add_attribute(mol, att, data, type="FLOAT_VECTOR")
    

def load(top, traj, name = 'oxDNA', world_scale = 0.01):
    
    # the scale of the oxDNA files seems to be based on nanometres rather than angstrongs 
    # like most structural biology files, so currently adjusting the world_scale to 
    # compensate
    dna_scale = world_scale * 10
    
    
    frames = read_oxdna(traj)
    topology = read_top(top)
    n_frames = frames.shape[0]
    
    # topology file will return a numeric numpy array, configured as follows:
    #
    # Strand assignment
    # | Base assignment
    # | |  3' Bonded base to the current base (index based on row)
    # | |  |  5' Bonded base to the current base (index based on row)
    # | |  | |
    # S B 3' 5'
    # S B 3' 5'
    # S B 3' 5'
    
    # to get pairs of indices which represent each distinct bond, which are needed for 
    # edge creation in Blender, take each bonded column and create a 'bond' with itself
    idx = np.array(list(range(topology.shape[0])))
    bond_3 = np.vstack((idx, topology[:, 2])).reshape((len(idx), 2))
    bond_5 = np.vstack((idx, topology[:, 3])).reshape((len(idx), 2))
    bonds = np.vstack((bond_3, bond_5))
    
    # drop where either bond is -1 (not bonded) from the bond indices
    mask = bonds == -1
    mask = np.logical_not(mask.any(axis=1)) 

    # creat toplogy object with positions of the first frame, and the bonds from the 
    # topology object
    mol = obj.create_object(
        name=name,
        collection=coll.mn(),
        locations=frames[0][:, 0:3] * dna_scale,
        bonds=bonds[mask, :]
    )
    
    # adding additional toplogy information from the topology and frames objects
    obj.add_attribute(mol, 'res_name', topology[:, 1], "INT")
    obj.add_attribute(mol, 'chain_id', topology[:, 0], "INT")
    add_attributes_to_dna_mol(mol, frames[0], dna_scale=dna_scale)
    
    # if the 'frames' file only contained one timepoint, return the object without creating
    # any kind of collection for storing multiple frames from a trajectory, and a None 
    # object in place of the frames collection
    if n_frames == 1:
        return mol, None
    
    # create a collection to store all of the frame objects that are part of the trajectory
    # they will contain all of the possible attributes which can be interpolated betewen 
    # frames such as position, base_vector, base_normal, velocity, angular_velocity
    collection = coll.frames(f"{name}_frames", parent=coll.data())
    for i, frame in enumerate(frames):
        fill_n = int(np.ceil(np.log10(n_frames)))
        frame_name = f"{name}_frame_{str(i).zfill(fill_n)}"
        frame_mol = obj.create_object(frame_name, collection, frame[:, 0:3] * dna_scale)
        add_attributes_to_dna_mol(frame_mol, frame, dna_scale)
    
    return mol, collection


def panel(layout_function, scene):
    col_main = layout_function.column(heading = "", align = False)
    col_main.label(text = "Import oxDNA File")
    row_import = col_main.row()
    row_import.prop(
        bpy.context.scene, 'MN_import_oxdna_name', 
        text = 'Name', 
        emboss = True
    )
    col_main.prop(bpy.context.scene, 'MN_import_oxdna_top')
    col_main.prop(bpy.context.scene, 'MN_import_oxdna_oxdna')
    row_import.operator('mn.import_oxdna', text = 'Load', icon = 'FILE_TICK')

class MN_OT_Import_Star_File(bpy.types.Operator):
    bl_idname = "mn.import_oxdna"
    bl_label = "Import oxDNA File"
    bl_description = "Will import the given file and toplogy."
    bl_options = {"REGISTER"}

    @classmethod
    def poll(cls, context):
        return True

    def execute(self, context):
        load(
            top=context.scene.MN_import_oxdna_top,
            traj=context.scene.MN_import_oxdna_oxdna, 
            name=context.scene.MN_import_oxdna_name
        )
        return {"FINISHED"}