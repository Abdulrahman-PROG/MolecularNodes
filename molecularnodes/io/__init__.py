from .parse import (
    CIF,
    BCIF,
    PDB,
    SDF,
    CellPack,
    StarFile,
    MMTF,
    MDAnalysisSession
)
from .wwpdb import fetch
from .local import load
