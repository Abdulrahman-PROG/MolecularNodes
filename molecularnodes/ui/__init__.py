from .ops import ops_ui
