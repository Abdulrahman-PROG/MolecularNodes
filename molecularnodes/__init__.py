# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTIBILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

import bpy

from .io import ops_io
from .io.parse.mda import _rejuvenate_universe, _sync_universe
from .io.parse.star import _rehydrate_ensembles
from .props import MolecularNodesObjectProperties
from .ui.node_menu import MN_add_node_menu, draw_node_menus
from .ui.ops import ops_ui
from .ui.panel import MN_PT_panel, change_style_menu, change_style_node_menu
from .utils import MN_OT_Install_Template

all_classes = ops_ui + ops_io + [MN_OT_Install_Template, MolecularNodesObjectProperties]

universe_funcs = [_sync_universe, _rejuvenate_universe]


def register():
    bpy.utils.register_class(MN_PT_panel)
    bpy.types.NODE_MT_add.append(MN_add_node_menu)

    # register all of the import operators
    for op in all_classes:
        bpy.utils.register_class(op)
    bpy.types.Object.mn = bpy.props.PointerProperty(type=MolecularNodesObjectProperties)
    bpy.types.Object.mn = bpy.props.PointerProperty(type=MolecularNodesObjectProperties)
    bpy.types.VIEW3D_MT_object_context_menu.prepend(change_style_menu)
    bpy.types.NODE_MT_context_menu.prepend(change_style_node_menu)
    for func in universe_funcs:
        try:
            bpy.app.handlers.load_post.append(func)
        except ValueError as e:
            print(f"Filaed to append {func}, error: {e}.")


def unregister():
    bpy.utils.unregister_class(MN_PT_panel)
    bpy.types.NODE_MT_add.remove(MN_add_node_menu)

    # unregister all of the import operator classes
    for op in all_classes:
        bpy.utils.unregister_class(op)

    try:
        bpy.types.NODE_MT_add.remove(MN_add_node_menu)
        bpy.types.VIEW3D_MT_object_context_menu.remove(change_style_menu)
        bpy.types.NODE_MT_context_menu.remove(change_style_node_menu)
        bpy.app.handlers.load_post.append(_rehydrate_ensembles)

        del bpy.types.Object.mn
        for func in universe_funcs:
            try:
                bpy.app.handlers.load_post.remove(func)
            except ValueError as e:
                print(f"Failed to remove {func}, error: {e}.")
    except RuntimeError:
        pass


# # can't register the add-on when these are uncommnted, but they do fix the issue
# # of having to call register() when running a script
# # unregister()
# # register()

# # # register won't be called when MN is run as a module
# bpy.app.handlers.load_post.append(_rejuvenate_universe)
# bpy.app.handlers.load_post.append(_rehydrate_ensembles)
# bpy.app.handlers.save_post.append(_sync_universe)
