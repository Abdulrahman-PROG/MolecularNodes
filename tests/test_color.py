import molecularnodes as mn
import numpy as np


def test_random_rgb(snapshot):
    n = 100
    colors = np.array(list(map(
        lambda x: mn.color.random_rgb(x),
        range(n)
    )))
    assert colors.tolist() == snapshot


def test_colos_from_atomic_numbers(snapshot):
    length = len(mn.color.iupac_colors_rgb)
    colors = mn.color.colors_from_elements(np.array(list(range(length))))
    assert colors.tolist() == snapshot
