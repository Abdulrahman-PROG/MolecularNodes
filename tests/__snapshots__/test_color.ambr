# serializer version: 1
# name: test_colos_from_atomic_numbers
  list([
    list([
      242,
      0,
      26,
      1,
    ]),
    list([
      255,
      255,
      255,
      1,
    ]),
    list([
      217,
      255,
      255,
      1,
    ]),
    list([
      204,
      128,
      255,
      1,
    ]),
    list([
      194,
      255,
      0,
      1,
    ]),
    list([
      255,
      181,
      181,
      1,
    ]),
    list([
      144,
      144,
      144,
      1,
    ]),
    list([
      48,
      80,
      248,
      1,
    ]),
    list([
      255,
      13,
      13,
      1,
    ]),
    list([
      144,
      224,
      80,
      1,
    ]),
    list([
      179,
      227,
      245,
      1,
    ]),
    list([
      171,
      92,
      242,
      1,
    ]),
    list([
      138,
      255,
      0,
      1,
    ]),
    list([
      191,
      166,
      166,
      1,
    ]),
    list([
      240,
      200,
      160,
      1,
    ]),
    list([
      255,
      128,
      0,
      1,
    ]),
    list([
      255,
      255,
      48,
      1,
    ]),
    list([
      31,
      240,
      31,
      1,
    ]),
    list([
      143,
      64,
      212,
      1,
    ]),
    list([
      128,
      209,
      227,
      1,
    ]),
    list([
      61,
      255,
      0,
      1,
    ]),
    list([
      230,
      230,
      230,
      1,
    ]),
    list([
      191,
      194,
      199,
      1,
    ]),
    list([
      166,
      166,
      171,
      1,
    ]),
    list([
      138,
      153,
      199,
      1,
    ]),
    list([
      156,
      122,
      199,
      1,
    ]),
    list([
      224,
      102,
      51,
      1,
    ]),
    list([
      199,
      138,
      138,
      1,
    ]),
    list([
      255,
      217,
      143,
      1,
    ]),
    list([
      200,
      128,
      51,
      1,
    ]),
    list([
      125,
      128,
      176,
      1,
    ]),
    list([
      194,
      143,
      143,
      1,
    ]),
    list([
      102,
      143,
      143,
      1,
    ]),
    list([
      189,
      128,
      227,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      166,
      41,
      41,
      1,
    ]),
    list([
      92,
      184,
      209,
      1,
    ]),
    list([
      112,
      46,
      176,
      1,
    ]),
    list([
      0,
      255,
      0,
      1,
    ]),
    list([
      148,
      255,
      255,
      1,
    ]),
    list([
      148,
      224,
      224,
      1,
    ]),
    list([
      115,
      194,
      201,
      1,
    ]),
    list([
      84,
      181,
      181,
      1,
    ]),
    list([
      59,
      158,
      158,
      1,
    ]),
    list([
      36,
      125,
      125,
      1,
    ]),
    list([
      10,
      125,
      140,
      1,
    ]),
    list([
      0,
      105,
      133,
      1,
    ]),
    list([
      192,
      192,
      192,
      1,
    ]),
    list([
      255,
      217,
      143,
      1,
    ]),
    list([
      166,
      117,
      115,
      1,
    ]),
    list([
      102,
      128,
      128,
      1,
    ]),
    list([
      158,
      99,
      181,
      1,
    ]),
    list([
      212,
      122,
      0,
      1,
    ]),
    list([
      148,
      0,
      148,
      1,
    ]),
    list([
      66,
      158,
      176,
      1,
    ]),
    list([
      87,
      23,
      143,
      1,
    ]),
    list([
      0,
      201,
      0,
      1,
    ]),
    list([
      112,
      212,
      255,
      1,
    ]),
    list([
      255,
      255,
      199,
      1,
    ]),
    list([
      217,
      255,
      199,
      1,
    ]),
    list([
      199,
      255,
      199,
      1,
    ]),
    list([
      163,
      255,
      199,
      1,
    ]),
    list([
      143,
      255,
      199,
      1,
    ]),
    list([
      97,
      255,
      199,
      1,
    ]),
    list([
      69,
      255,
      199,
      1,
    ]),
    list([
      48,
      255,
      199,
      1,
    ]),
    list([
      31,
      255,
      199,
      1,
    ]),
    list([
      0,
      255,
      156,
      1,
    ]),
    list([
      0,
      230,
      117,
      1,
    ]),
    list([
      0,
      212,
      82,
      1,
    ]),
    list([
      0,
      191,
      56,
      1,
    ]),
    list([
      0,
      171,
      36,
      1,
    ]),
    list([
      77,
      194,
      255,
      1,
    ]),
    list([
      77,
      166,
      255,
      1,
    ]),
    list([
      33,
      148,
      214,
      1,
    ]),
    list([
      38,
      125,
      171,
      1,
    ]),
    list([
      38,
      102,
      150,
      1,
    ]),
    list([
      23,
      84,
      135,
      1,
    ]),
    list([
      208,
      208,
      224,
      1,
    ]),
    list([
      255,
      209,
      35,
      1,
    ]),
    list([
      184,
      184,
      208,
      1,
    ]),
    list([
      166,
      84,
      77,
      1,
    ]),
    list([
      87,
      89,
      97,
      1,
    ]),
    list([
      158,
      79,
      181,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      255,
      161,
      0,
      1,
    ]),
    list([
      204,
      0,
      89,
      1,
    ]),
    list([
      209,
      0,
      79,
      1,
    ]),
    list([
      217,
      0,
      69,
      1,
    ]),
    list([
      224,
      0,
      56,
      1,
    ]),
    list([
      230,
      0,
      46,
      1,
    ]),
    list([
      235,
      0,
      38,
      1,
    ]),
    list([
      240,
      0,
      33,
      1,
    ]),
    list([
      241,
      0,
      30,
      1,
    ]),
    list([
      242,
      0,
      26,
      1,
    ]),
    list([
      242,
      0,
      26,
      1,
    ]),
    list([
      242,
      0,
      26,
      1,
    ]),
    list([
      242,
      0,
      26,
      1,
    ]),
    list([
      242,
      0,
      26,
      1,
    ]),
    list([
      242,
      0,
      26,
      1,
    ]),
  ])
# ---
# name: test_random_rgb
  list([
    list([
      0.7248194412335895,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.7469690230437155,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.48662129695896145,
      1.0,
    ]),
    list([
      0.6346618739753529,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.6401815015561884,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.4860431187176583,
      0.84,
      1.0,
    ]),
    list([
      0.7248194412335895,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.38736163728049217,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.6670871249825792,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.7334611905072618,
      1.0,
    ]),
    list([
      0.36,
      0.634360527293049,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.7028531141082777,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.7667635546230185,
      1.0,
    ]),
    list([
      0.5740555438594361,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.6676661885807683,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.4601026323298859,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.44118559175253647,
      1.0,
    ]),
    list([
      0.36,
      0.7766863400280194,
      0.84,
      1.0,
    ]),
    list([
      0.7979571936003222,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.39012238118477865,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.6317577326173801,
      1.0,
    ]),
    list([
      0.84,
      0.8350545019196102,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.4803569861262539,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.5763880753172774,
      1.0,
    ]),
    list([
      0.4915478049415251,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.4856514308843116,
      1.0,
    ]),
    list([
      0.5921988467230564,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.4123280064608166,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.6853166501073495,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.7014171250223065,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.7274450939352655,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.3953613525061693,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.5829747374975158,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.6374541413658052,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.7566660971316124,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.6997595357457919,
      0.84,
      1.0,
    ]),
    list([
      0.373342500864678,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.40417313449337605,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.4383157592054773,
      0.84,
      1.0,
    ]),
    list([
      0.7156284157348304,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.7207873230347368,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.4973395871878218,
      1.0,
    ]),
    list([
      0.36,
      0.4384508204412947,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.47102929729165455,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.576583332252942,
      1.0,
    ]),
    list([
      0.5369988065738183,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.6817879398168343,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.41331721678430206,
      1.0,
    ]),
    list([
      0.36,
      0.7013054541609705,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.5526258518671248,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.8329053180247746,
      1.0,
    ]),
    list([
      0.6186688433971095,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.42233800789152925,
      1.0,
    ]),
    list([
      0.36,
      0.5026325471905375,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.6087082390419132,
      1.0,
    ]),
    list([
      0.84,
      0.6201787802957266,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.45815273848816085,
      1.0,
    ]),
    list([
      0.84,
      0.4825892551178214,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.6074193649902047,
      0.84,
      1.0,
    ]),
    list([
      0.6743458324397,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.4335165463667209,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.8241238156207719,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.5673842902722976,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.6820009684516497,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.7713357952205155,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.5944627180089955,
      1.0,
    ]),
    list([
      0.84,
      0.563929709930129,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.575406019794559,
      0.36,
      1.0,
    ]),
    list([
      0.5760065345699711,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.4106139429467711,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.6198645947792483,
      1.0,
    ]),
    list([
      0.38759036403619473,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.5714687658307149,
      0.36,
      1.0,
    ]),
    list([
      0.5141314361097951,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.7639820778474273,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.6984339863376261,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.4672031263573643,
      1.0,
    ]),
    list([
      0.7414522630561735,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.785697058889694,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.7785681460798484,
      0.36,
      1.0,
    ]),
    list([
      0.5381055938113568,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.8228916083416866,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.7767951915849651,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.8289297690251884,
      1.0,
    ]),
    list([
      0.5468096555526074,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.7451845257001914,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.7195946955690216,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.84,
      0.7803498259573831,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.5447679657065977,
      1.0,
    ]),
    list([
      0.84,
      0.5935411141167872,
      0.36,
      1.0,
    ]),
    list([
      0.7334291114273717,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.84,
      0.6005391009223267,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.6126084891455414,
      1.0,
    ]),
    list([
      0.84,
      0.36,
      0.612451125043922,
      1.0,
    ]),
    list([
      0.36,
      0.7052115258144966,
      0.84,
      1.0,
    ]),
    list([
      0.6292650575321505,
      0.36,
      0.84,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.4673116131590971,
      1.0,
    ]),
    list([
      0.7597327259524387,
      0.84,
      0.36,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.42642683547663984,
      1.0,
    ]),
    list([
      0.36,
      0.84,
      0.5634568558377591,
      1.0,
    ]),
  ])
# ---
